File: examples/text2/src/main.rs

use std::fs::File;
use std::rc::Rc;

use simplelog::{WriteLogger, LevelFilter, ConfigBuilder};

use green_moon_2d::{GMEngine, GMSceneT, GMContext, GMEventCode};
use green_moon_2d::bitmap_text::{GMBitmapText};
use green_moon_2d::util::GMAlign;
use green_moon_2d::object_manager::GMObjectManager;
use green_moon_2d::target::GMTarget;
use green_moon_2d::value::GMValue;

#[derive(Debug)]
struct TextScene2 {
    object_manager: GMObjectManager,
    target: GMTarget,
}

impl TextScene2 {
    fn new(engine: &GMEngine) -> Self {
        // Access to resources
        let resources = engine.get_resources();

        // Crate object manager
        let mut object_manager = GMObjectManager::new();

        // Get first font
        let font1 = resources.get_font("font_cuddly");

        // Set up title text:
        let mut text = GMBitmapText::new(font1, (512.0, 100.0), "TEXT 2");
        text.set_align(GMAlign::BottomCenter);

        object_manager.add_draw_object("title", text, 0, 0);

        // Set up description:
        text = GMBitmapText::new(font1, (512.0, 200.0), "PRESS 1 - 9");
        text.set_align(GMAlign::BottomCenter);

        object_manager.add_draw_object("description1", text, 0, 0);

        text = GMBitmapText::new(font1, (512.0, 300.0), "TO CHANGE ALIGNMENT");
        text.set_align(GMAlign::BottomCenter);

        object_manager.add_draw_object("description2", text, 0, 0);

        text = GMBitmapText::new(font1, (512.0, 350.0), "PRESS H FOR ORIENTATION");
        text.set_align(GMAlign::BottomCenter);

        object_manager.add_draw_object("description3", text, 0, 0);

        text = GMBitmapText::new(font1, (512.0, 450.0), "--- DEMO TEXT ---");
        text.set_align(GMAlign::BottomCenter);

        // Draw index is 1, so it's drawn over the other text
        object_manager.add_draw_object("demo_text", text, 0, 1);

        Self {
            object_manager,
            target: GMTarget::Single("demo_text".to_string()),
        }
    }

    fn change_alignment(&mut self, align: GMAlign, context: &mut GMContext) {
        self.object_manager.send_custom_message1(&self.target, "set_align2", GMValue::Any(Rc::new(align)), context);
    }
}

impl GMSceneT for TextScene2 {
    fn update(&mut self, context: &mut GMContext) {
        if context.event(GMEventCode::KeyESCUp) ||
           context.event(GMEventCode::Quit) ||
           context.event(GMEventCode::WindowClose) {
            context.quit();
        }

        if context.event(GMEventCode::Key1Up) {
            self.change_alignment(GMAlign::TopLeft, context);
        }

        if context.event(GMEventCode::Key2Up) {
            self.change_alignment(GMAlign::TopCenter, context);
        }

        if context.event(GMEventCode::Key3Up) {
            self.change_alignment(GMAlign::TopRight, context);
        }

        if context.event(GMEventCode::Key4Up) {
            self.change_alignment(GMAlign::MiddleLeft, context);
        }

        if context.event(GMEventCode::Key5Up) {
            self.change_alignment(GMAlign::MiddleCenter, context);
        }

        if context.event(GMEventCode::Key6Up) {
            self.change_alignment(GMAlign::MiddleRight, context);
        }

        if context.event(GMEventCode::Key7Up) {
            self.change_alignment(GMAlign::BottomLeft, context);
        }

        if context.event(GMEventCode::Key8Up) {
            self.change_alignment(GMAlign::BottomCenter, context);
        }

        if context.event(GMEventCode::Key9Up) {
            self.change_alignment(GMAlign::BottomRight, context);
        }

        if context.event(GMEventCode::KeyHUp) {
            self.object_manager.send_custom_message0(&self.target, "toggle_horizontal2", context);
        }

        self.object_manager.update(context);
    }

    fn draw(&self, context: &mut GMContext) {
        context.clear_black();

        self.object_manager.draw(context);
    }
}

fn main() {
    let log_config = ConfigBuilder::new().build();
    let _simple_log = WriteLogger::init(LevelFilter::Debug, log_config, File::create("text2.log").expect("Could not create log file"));

    let mut engine = GMEngine::new();
    engine.load_configuration_and_init("config.json");

    let text2_scene = TextScene2::new(&engine);

    engine.add_scene("text2_scene", text2_scene);
    engine.run();
}

File: examples/text2/main.py
# This file is part of GreenMoon2D, a 2D game engine for Python
# written by Willi Kappler, MIT license.
#
# See: https://github.com/willi-kappler/green_moon_2d

from typing import override
import sys

sys.path.append("../../src/")

import green_moon_2d.gm_engine as gme
from green_moon_2d.gm_scene import GMScene
from green_moon_2d.gm_text import GMText
from green_moon_2d.gm_math import GMAlignment
from green_moon_2d.gm_object import GMObjectManager


class TestScene(GMScene):
    def __init__(self):
        super().__init__("TestScene2")

    @override
    def enter(self):
        if not self.initialized:
            resources = gme.GMGlobalResources
            font = resources.get_font("font_cuddly")

            center = GMAlignment.TOP_CENTER

            self.om = GMObjectManager()

            self.om.add(GMText("text1", "TEXT 2", (512.0, 100.0), font, center))
            self.om.add(GMText("text2", "PRESS 1 - 9", (512.0, 200.0), font, center))
            self.om.add(GMText("text3", "TO CHANGE ALIGNMENT", (512.0, 250.0), font, center))
            self.om.add(GMText("text4", "PRESS H FOR ORIENTATION", (512.0, 350.0), font, center))
            self.demo_text = GMText("text5", "DEMO TEXT", (512.0, 450.0), font, center)
            self.om.add(self.demo_text)

            # Draw text5 after (= on top of) all other texts since it may overlap.
            self.demo_text.draw_order = 1
            self.om.sort_draw()

            self.initialized = True

    @override
    def update(self):
        keys_up = gme.GMGlobalContext.keys_up
        keys_down = gme.GMGlobalContext.keys_down

        if "ESC" in keys_down:
            gme.GMGlobalContext.quit_game = True

        if "1" in keys_down:
            self.demo_text.set_alignment(GMAlignment.TOP_LEFT)

        if "2" in keys_down:
            self.demo_text.set_alignment(GMAlignment.TOP_CENTER)

        if "3" in keys_down:
            self.demo_text.set_alignment(GMAlignment.TOP_RIGHT)

        if "4" in keys_down:
            self.demo_text.set_alignment(GMAlignment.MID_LEFT)

        if "5" in keys_down:
            self.demo_text.set_alignment(GMAlignment.MID_CENTER)

        if "6" in keys_down:
            self.demo_text.set_alignment(GMAlignment.MID_RIGHT)

        if "7" in keys_down:
            self.demo_text.set_alignment(GMAlignment.BTM_LEFT)

        if "8" in keys_down:
            self.demo_text.set_alignment(GMAlignment.BTM_CENTER)

        if "9" in keys_down:
            self.demo_text.set_alignment(GMAlignment.BTM_RIGHT)

        if "H" in keys_up:
            self.demo_text.toggle_orientation()

    @override
    def draw(self):
        self.om.draw()


def main():
    engine = gme.GMEngine("config.json", "TestScene2")

    scene1 = TestScene()
    gme.GMGlobalScenes.add_scene(scene1)
    engine.run()


if __name__ == "__main__":
    main()

